File: practica parte 2/Ejercicio 3: Cliente y ArchivoCliente/src/cliente.java
import java.io.*;
import java.util.ArrayList;
import java.util.List;

class Cliente implements Serializable {
    private int id;
    private String nombre;
    private int telefono;
    
    public Cliente(int id, String nombre, int telefono) {
        this.id = id;
        this.nombre = nombre;
        this.telefono = telefono;
    }
    
    public int getId() {
        return id;
    }
    
    public String getNombre() {
        return nombre;
    }
    
    public int getTelefono() {
        return telefono;
    }
}

class ArchivoCliente {
    private String nomA;
    
    public ArchivoCliente(String nombreArchivo) {
        this.nomA = nombreArchivo;
    }
    
    public void crearArchivo() {
        try (ObjectOutputStream oos = new ObjectOutputStream(new FileOutputStream(nomA))) {
            oos.writeObject(new ArrayList<Cliente>());
        } catch (IOException e) {
            System.out.println("Error al crear archivo: " + e.getMessage());
        }
    }
    
    public void guardaCliente(Cliente c) {
        try (ObjectInputStream ois = new ObjectInputStream(new FileInputStream(nomA))) {
            List<Cliente> clientes = (List<Cliente>) ois.readObject();
            clientes.add(c);
            
            try (ObjectOutputStream oos = new ObjectOutputStream(new FileOutputStream(nomA))) {
                oos.writeObject(clientes);
            }
        } catch (IOException | ClassNotFoundException ex) {
            System.out.println("Error al guardar cliente: " + ex.getMessage());
        }
    }
    
    public Cliente buscarCliente(int c) {
        try (ObjectInputStream ois = new ObjectInputStream(new FileInputStream(nomA))) {
            List<Cliente> clientes = (List<Cliente>) ois.readObject();
            
            for (Cliente cli : clientes) {
                if (cli.getId() == c) {
                    return cli;
                }
            }
        } catch (IOException | ClassNotFoundException e) {
            System.out.println("Error al buscar cliente: " + e.getMessage());
        }
        return null;
    }
    
    public Cliente buscarCelularCliente(int c) {
        try (ObjectInputStream ois = new ObjectInputStream(new FileInputStream(nomA))) {
            List<Cliente> clientes = (List<Cliente>) ois.readObject();
            
            for (Cliente cli : clientes) {
                if (cli.getTelefono() == c) {
                    return cli;
                }
            }
        } catch (IOException | ClassNotFoundException e) {
            System.out.println("Error al buscar cliente por teléfono: " + e.getMessage());
        }
        return null;
    }

    public static void main(String[] args) {
        ArchivoCliente archivo = new ArchivoCliente("clientes.dat");
        archivo.crearArchivo();
        
        archivo.guardaCliente(new Cliente(1, "Carlos Ruiz", 7654321));
        archivo.guardaCliente(new Cliente(2, "Ana Xambrana", 7123456));
        
        Cliente cli = archivo.buscarCliente(1);
        if (cli != null) {
            System.out.println("Cliente encontrado: " + cli.getNombre() + ", Tel: " + cli.getTelefono());
        }
        
        Cliente cliCel = archivo.buscarCelularCliente(7123456);
        if (cliCel != null) {
            System.out.println("Cliente por teléfono: " + cliCel.getNombre());
        }
    }
}

File: practica parte 2/Ejercicio 3: Cliente y ArchivoCliente/cliente.py
import pickle

class Cliente:
    def __init__(self, id_cliente, nombre, telefono):
        self.id = id_cliente
        self.nombre = nombre
        self.telefono = telefono
    
    def get_id(self):
        return self.id
    
    def get_nombre(self):
        return self.nombre
    
    def get_telefono(self):
        return self.telefono

class ArchivoCliente:
    def __init__(self, nombre_archivo):
        self.nomA = nombre_archivo
    
    def crear_archivo(self):
        try:
            with open(self.nomA, 'wb') as f:
                pickle.dump([], f)
        except IOError:
            print("Error al crear archivo")
    
    def guarda_cliente(self, cliente):
        try:
            with open(self.nomA, 'rb') as f:
                clientes = pickle.load(f)
            
            clientes.append(cliente)
            
            with open(self.nomA, 'wb') as f:
                pickle.dump(clientes, f)
        except IOError:
            print("Error al guardar cliente")
    
    def buscar_cliente(self, id_cliente):
        try:
            with open(self.nomA, 'rb') as f:
                clientes = pickle.load(f)
            
            for cli in clientes:
                if cli.get_id() == id_cliente:
                    return cli
            return None
        except IOError:
            print("Error al buscar cliente")
            return None
    
    def buscar_celular_cliente(self, telefono):
        try:
            with open(self.nomA, 'rb') as f:
                clientes = pickle.load(f)
            
            for cli in clientes:
                if cli.get_telefono() == telefono:
                    return cli
            return None
        except IOError:
            print("rror al buscar cliente por teléfono")
            return None

archivo = ArchivoCliente("clientes.dat")
archivo.crear_archivo()

archivo.guarda_cliente(Cliente(1, "Carlos Ruiz", 7654321))
archivo.guarda_cliente(Cliente(2, "Ana Zambrano", 7123456))

cli = archivo.buscar_cliente(1)
if cli:
    print(f"Cliente encontrado: {cli.get_nombre()}, Tel: {cli.get_telefono()}")

cli_cel = archivo.buscar_celular_cliente(7123456)
if cli_cel:
    print(f"xliente por teléfono: {cli_cel.get_nombre()}")